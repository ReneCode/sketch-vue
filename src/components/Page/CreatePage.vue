<template>
  <v-container>
    <v-layout row>
      <v-flex xs12 sm6 offset-sm3>
        <v-card>
          <v-card-text>
            <v-container>
              <form @submit.prevent="onCreatePage">
                <v-layout row>
                  <v-flex xs12>
                    <v-text-field name="name" label="Name" id="name" v-model="name" type="text" required>
                    </v-text-field>
                    <v-text-field name="description" label="Description" id="description" v-model="description" type="text">
                    </v-text-field>
                  </v-flex>
                </v-layout>
                <v-layout>
                  <v-flex x12>
                    <v-btn type="submit" :disabled="!validName">Create page</v-btn>
                  </v-flex>
                </v-layout>
              </form>
            </v-container>
          </v-card-text>
        </v-card>
      </v-flex>
    </v-layout>
  </v-container>
</template>

<script>
export default {
  props: ['projectId'],

  data() {
    return {
      name: '',
      description: ''
    }
  },

  computed: {
    validName() {
      return this.name.length > 0;
    }
  },

  methods: {
    onCreatePage() {
      const page = {
        projectId: this.projectId,
        name: this.name,
        description: this.description
      }
      this.$store.dispatch('createPage', page)
        .then(pageId => {
          this.$router.push('/projects/' + this.projectId + '/pages/' + pageId)
        })
    }

  }
}
</script>
