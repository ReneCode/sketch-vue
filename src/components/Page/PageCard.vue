<template>
  <v-card class="card">
    <v-card-title primary-title>
      <h4>{{page.name}}</h4>
    </v-card-title>
    <v-card-text>{{page.description}}</v-card-text>
    </v-card-actions>
    <v-btn class="hover-show" @click="openPage(page.id)">
      <v-icon left>arrow_forward</v-icon>
      Show Page
    </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script>
export default {
  props: ["page", 'projectId'],
  methods: {
    openPage(id) {
      this.$router.push('/projects/' + this.projectId + '/pages/' + id);
    }
  }
};
</script>

<style>
.card {
  width: 300px;
  height: 100%;
  padding-bottom: 40px;
}
</style>
