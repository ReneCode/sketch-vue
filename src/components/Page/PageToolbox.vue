<template>
  <div class="toolbox">
    <v-btn fab small class="green" @click="onColor('green')"></v-btn>
    <v-btn fab small class="red " @click="onColor('red')"></v-btn>
    <v-btn fab small class="black mr-3" @click="onColor('black')"></v-btn>

    <v-btn fab :color="iaMode === 'select'?'primary':'accent'" @click="onSetInteractionMode('select')">
      <v-icon class="flipy" dark>call_made</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'panning'?'primary':'accent'" @click="onSetInteractionMode('panning')">
      <v-icon dark>pan_tool</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'rectangle'?'primary':'accent'" @click="onSetInteractionMode('rectangle')">
      <v-icon dark>check_box_outline_blank</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'circle'?'primary':'accent'" @click="onSetInteractionMode('circle')">
      <v-icon dark>radio_button_unchecked</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'polygon'?'primary':'accent'" @click="onSetInteractionMode('polygon')">
      <v-icon dark>details</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'freehand'?'primary':'accent'" @click="onSetInteractionMode('freehand')">
      <v-icon dark>edit</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'delete'?'primary':'accent'" @click="onSetInteractionMode('delete')">
      <v-icon dark>delete</v-icon>
    </v-btn>
    <v-btn fab :color="iaMode === 'image'?'primary':'accent'" @click="onSetInteractionMode('image')">
      <v-icon dark>add_a_photo</v-icon>
    </v-btn>
<!-- 
    <v-btn fab color="primary" @click="onShowDialogUploadPicture()">
      <v-icon dark>add_a_photo</v-icon>
    </v-btn> -->

    <v-btn fab color="primary" :disabled="!undoRedoList.canUndo" @click="onUndo" class="ml-4">
      <v-icon dark>replay</v-icon>
    </v-btn>
    <v-btn fab color="primary" :disabled="!undoRedoList.canRedo" @click="onRedo">
      <v-icon class="flipy" dark>replay</v-icon>
    </v-btn>
  </div>
</template>

<script>
import undoRedoList from "../../store/modules/undo-redo-list";
import selectionList from "../../store/selection-list";

export default {
  data() {
    return {
      undoRedoList: undoRedoList
    };
  },

  computed: {

    iaMode() {
      return this.$store.getters.interactionMode;
    },

    loading() {
      return this.$store.getters.loading;
    }
  },

  mounted() {
    this.onSetInteractionMode("select");
  },

  methods: {
    onSetInteractionMode(mode) {
      const payload = {
        mode: mode,
        projectId: this.projectId,
        pageId: this.pageId,
        fileInput: this.$refs.fileInput
      };
      this.$store.commit("setInteractionMode", payload);
    },
    onUndo() {
      selectionList.clear();
      undoRedoList.undo();
    },
    onRedo() {
      selectionList.clear();
      undoRedoList.redo();
    },
    onColor(color) {
      this.$store.commit("setColor", color);
    }
  }
};
</script>

<style>
.flipy {
  -moz-transform: scaleX(-1);
  -o-transform: scaleX(-1);
  -webkit-transform: scaleX(-1);
  transform: scaleX(-1);
  filter: FlipH;
  -ms-filter: "FlipH";
}

.toolbox {
  top: 0px;
  right: 0px;
  position: absolute;
}

code {
  font-size: 11px;
  line-height: 1;
  color: #444;
}
</style>
