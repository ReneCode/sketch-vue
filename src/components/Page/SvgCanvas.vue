<template>
  <svg ref="svg" width="100%" height="100%">
    <g :transform="svgTransform">
      <!-- <image xlink:href="http://lorempixel.com/400/200/technics" x="100" y="80"/> -->
      <svg-item v-for="(item,index) in allItems" :key="index" :item="item" :iid="item.id" :class="item.selected? 'item-selected': 'item-normal'"></svg-item>
      <svg-item v-for="(item,index) in tmpItems" :key="index" :item="item" class="tmp" :iid="item.id"></svg-item>
    </g>
  </svg>
</template>

<script>
import Svg from "../../svg";
import SvgItem from "./SvgItem";

export default {
  name: "svg-canvas",
  props: ["allItems", "tmpItems"],
  components: {
    SvgItem
  },
  data() {
    return {
      svg: {}
    };
  },

  computed: {
    svgTransform() {
      if (this.svg) {
        return this.svg.getSvgTransformString();
      } else {
        return "";
      }
    }
  },

  created() {
    this.svg = new Svg();
  },

  mounted() {
    this.svg.init(this.$refs.svg);
  },

  beforeDestroy() {
    this.svg.exit();
  },

  methods: {}
};
</script>

<style>
svg {
  background-color: #f0f0f0;
}

.item-normal {
  stroke: #630;
  stroke-width: 2px;
  opacity: 0.9;
}

.item-selected {
  stroke: #222;
  stroke-width: 2px;
  stroke-dasharray: 5;
  /* cursor: pointer; */
  opacity: 0.8;
  animation: dash-rotate 0.5s linear infinite;
}

@keyframes dash-rotate {
  from {
    stroke-dashoffset: 10;
  }
  to {
    stroke-dashoffset: 0;
  }
}

.tmp {
  /* fill: #bbf; */
  stroke: #33c;
  cursor: pointer;
  opacity: 0.3;
}
</style>
