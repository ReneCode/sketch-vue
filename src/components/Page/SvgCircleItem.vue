<template>
  <circle :cx="item.svg.x" :cy="item.svg.y" :r="item.svg.radius" :style="style">
  </circle>
</template>

<script>
export default {
  name: 'svg-circle-item',
  props: ['item'],
  computed: {
    style() {
      let style = {
      };
      if (this.item.svg.stroke) {
        style.stroke = this.item.svg.stroke;
      }
      if (this.item.svg.strokeWidth) {
        style.strokeWidth = this.item.svg.strokeWidth;
      }
      if (this.item.svg.fill) {
        style.fill = this.item.svg.fill;
      }
      return style;
    }
  }
}
</script>

<style>
circle {
  stroke: #630;
  stroke-width: 2px;
  fill: none;
}
</style>
