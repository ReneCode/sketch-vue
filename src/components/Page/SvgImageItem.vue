<template>
  <image :x="item.svg.x" 
         :y="item.svg.y" 
         :xlink:href="item.svg.imageUrl">
  </image>
</template>

<script>
export default {
  name: "svg-image-item",
  props: ["item"]
};
</script>

<style>

</style>
