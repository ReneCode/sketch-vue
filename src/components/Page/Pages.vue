<template>
  <v-container>
    <v-layout row class="mb-3">
      <v-flex xs12 class="text-xs-left">
        <v-btn large @click="newPage">Create new Page</v-btn>
      </v-flex>
    </v-layout>
    <v-layout row wrap v-if="!loading">
      <v-flex v-for="page in pages" :key="page.name">
        <app-page-card :projectId="projectId" :page="page"></app-page-card>
      </v-flex>
    </v-layout>
  </v-container>
</template>

<script>

import PageCard from './PageCard'

export default {
  props: ['projectId'],

  components: {
    'app-page-card': PageCard
  },

  computed: {
    pages() {
      return this.$store.getters.loadedPages;
    },

    loading() {
      return this.$store.getters.loading;
    }
  },

  mounted() {
    this.$store.dispatch('loadPages', this.projectId);
  },

  methods: {
    newPage() {
      this.$router.push('/projects/' + this.projectId + '/pages/new');
    }
  }
}
</script>

<style>
</style>
