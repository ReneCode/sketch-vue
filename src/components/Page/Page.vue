<template>
  <v-container fluid>
    <div class="toolbox">
      <v-btn fab small class="green" @click="onColor('green')"></v-btn>
      <v-btn fab small class="red " @click="onColor('red')"></v-btn>
      <v-btn fab small class="black mr-3" @click="onColor('black')"></v-btn>

      <v-btn fab :color="iaMode === 'select'?'primary':'accent'" @click="onSetInteractionMode('select')">
        <v-icon class="flipy" dark>call_made</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'panning'?'primary':'accent'" @click="onSetInteractionMode('panning')">
        <v-icon dark>pan_tool</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'rectangle'?'primary':'accent'" @click="onSetInteractionMode('rectangle')">
        <v-icon dark>check_box_outline_blank</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'circle'?'primary':'accent'" @click="onSetInteractionMode('circle')">
        <v-icon dark>radio_button_unchecked</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'polygon'?'primary':'accent'" @click="onSetInteractionMode('polygon')">
        <v-icon dark>details</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'freehand'?'primary':'accent'" @click="onSetInteractionMode('freehand')">
        <v-icon dark>edit</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'delete'?'primary':'accent'" @click="onSetInteractionMode('delete')">
        <v-icon dark>delete</v-icon>
      </v-btn>
      <v-btn fab :color="iaMode === 'image'?'primary':'accent'" @click="onSetInteractionMode('image')">
        <v-icon dark>add_a_photo</v-icon>
      </v-btn>
<!-- 
      <v-btn fab color="primary" @click="onShowDialogUploadPicture()">
        <v-icon dark>add_a_photo</v-icon>
      </v-btn> -->

      <v-btn fab color="primary" :disabled="!undoRedoList.canUndo" @click="onUndo" class="ml-4">
        <v-icon dark>replay</v-icon>
      </v-btn>
      <v-btn fab color="primary" :disabled="!undoRedoList.canRedo" @click="onRedo">
        <v-icon class="flipy" dark>replay</v-icon>
      </v-btn>
    </div>
    <div class="graphic">
      <svg-canvas :tmpItems="tmpItems" :allItems="allItems"></svg-canvas>
    </div>

    <app-upload-picture :show="showDialogUploadPicture" @upload="onUploadPicture"  @close="onCloseUploadPicture"></app-upload-picture>

  <!-- That fileInput is used for selecting an local image on "iaImage" -->
    <input ref="fileInput" type="file" style="display:none" accept="image/*">

  </v-container>
</template>

<script>
import undoRedoList from "../../store/modules/undo-redo-list";
import selectionList from "../../store/selection-list";
import temporaryItemList from "../../store/temporary-item-list";
import interaction from "../../interaction";
import SvgCanvas from "./SvgCanvas";

export default {
  props: ["projectId", "pageId"],
  components: {
    "svg-canvas": SvgCanvas
  },
  data() {
    return {
      tmpItems: temporaryItemList.getItems(),
      selectedItems: selectionList.getItems(),
      showDialogUploadPicture: false,
      iaList: interaction.getIaList(),
      undoRedoList: undoRedoList
    };
  },

  computed: {
    allItems: function() {
      return this.loadedGraphics
        .map(item => {
          // do not show items, that are in temporaryData
          let selectedItem = this.selectedItems.find(
            selItem => selItem.id === item.id
          );
          if (selectedItem) {
            // item is selected
            return selectedItem;
          } else {
            let tempItem = this.tmpItems.find(
              tmpItem => tmpItem.id === item.id
            );
            if (tempItem) {
              // item is temporary - will be removed (.filter())
              return null;
            }
            // quite normal item
            return item;
          }
        })
        .filter(item => item !== null);
    },

    loadedGraphics() {
      return this.$store.getters.loadedGraphics;
    },

    urList() {
      return this.undoRedoList.getList().map(ur => {
        if (!ur.ref) {
          return ur;
        } else {
          return {
            old: ur.oldData,
            new: ur.newData
          };
        }
      });
    },

    iaMode() {
      return this.$store.getters.interactionMode;
    },

    loading() {
      return this.$store.getters.loading;
    }
  },

  mounted() {
    // projectId and pageId are mounted via routing
    this.$store.dispatch("setCurrentProjectId", this.projectId);
    this.$store.dispatch("setCurrentPageId", this.pageId);

    this.$store.dispatch("loadGraphics");
    this.onSetInteractionMode("select");
  },

  methods: {
    onSetInteractionMode(mode) {
      const payload = {
        mode: mode,
        fileInput: this.$refs.fileInput
      };
      this.$store.commit("setInteractionMode", payload);
    },
    onUndo() {
      selectionList.clear();
      undoRedoList.undo();
    },
    onRedo() {
      selectionList.clear();
      undoRedoList.redo();
    },
    onColor(color) {
      this.$store.commit("setColor", color);
    },
    onShowDialogUploadPicture() {
      this.showDialogUploadPicture = true;
    },
    onUploadPicture(file) {
      console.log("F:", file);
      this.showDialogUploadPicture = false;
    },
    onCloseUploadPicture() {
      this.showDialogUploadPicture = false;
    }
  }
};
</script>

<style>
.flipy {
  -moz-transform: scaleX(-1);
  -o-transform: scaleX(-1);
  -webkit-transform: scaleX(-1);
  transform: scaleX(-1);
  filter: FlipH;
  -ms-filter: "FlipH";
}

.graphic {
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  position: absolute;
}

.toolbox {
  top: 0px;
  right: 0px;
  position: absolute;
}

code {
  font-size: 11px;
  line-height: 1;
  color: #444;
}
</style>
