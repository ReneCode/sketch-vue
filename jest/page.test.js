import Vue from 'vue';
import Vuex from 'vuex';
import { mount } from 'vue-test-utils'
import Page from '../src/components/Page/Page'
import ItemCircle from '../src/models/item-circle'
import Point from '../src/models/point'

Vue.use(Vuex);

const createStore = (loadedGraphics = []) => {
  return new Vuex.Store({
    state: {},
    getters: {
      loadedGraphics: () => {
        return loadedGraphics;
      }
    },
    mutations: {
      setInteractionMode: () => { }
    },
    actions: {
      loadGraphics: () => { },
      setCurrentPageId: () => { },
      setCurrentProjectId: () => { }
    }
  })
}

describe('Page', () => {
  let cmp
  let store

  it('renders empty page', () => {
    store = createStore();
    cmp = mount(Page, { store })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('renders svg with allItems = circle', () => {
    const item = new ItemCircle(new Point(50, 20), 40);
    store = createStore([item]);
    cmp = mount(Page, { store })
    expect(cmp.html()).toMatchSnapshot()
  })
})
