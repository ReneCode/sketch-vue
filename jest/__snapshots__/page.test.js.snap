// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders empty page 1`] = `
<v-container fluid="">
    <div class="toolbox">
        <v-btn fab="" small="" class="green"></v-btn>
        <v-btn fab="" small="" class="red "></v-btn>
        <v-btn fab="" small="" class="black mr-3"></v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="" class="flipy">call_made</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">pan_tool</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">check_box_outline_blank</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">radio_button_unchecked</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">details</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">edit</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">delete</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">add_a_photo</v-icon>
        </v-btn>
        <v-btn fab="" color="primary" disabled="true" class="ml-4">
            <v-icon dark="">replay</v-icon>
        </v-btn>
        <v-btn fab="" color="primary" disabled="true">
            <v-icon dark="" class="flipy">replay</v-icon>
        </v-btn>
    </div>
    <div class="graphic"><svg width="100%" height="100%"><g transform="translate(0,0)scale(1)"> </g></svg></div>
    <app-upload-picture></app-upload-picture> <input type="file" accept="image/*" style="display: none;"></v-container>
`;

exports[`Page renders svg with allItems = circle 1`] = `
<v-container fluid="">
    <div class="toolbox">
        <v-btn fab="" small="" class="green"></v-btn>
        <v-btn fab="" small="" class="red "></v-btn>
        <v-btn fab="" small="" class="black mr-3"></v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="" class="flipy">call_made</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">pan_tool</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">check_box_outline_blank</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">radio_button_unchecked</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">details</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">edit</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">delete</v-icon>
        </v-btn>
        <v-btn fab="" color="accent">
            <v-icon dark="">add_a_photo</v-icon>
        </v-btn>
        <v-btn fab="" color="primary" disabled="true" class="ml-4">
            <v-icon dark="">replay</v-icon>
        </v-btn>
        <v-btn fab="" color="primary" disabled="true">
            <v-icon dark="" class="flipy">replay</v-icon>
        </v-btn>
    </div>
    <div class="graphic"><svg width="100%" height="100%"><g transform="translate(0,0)scale(1)"><circle cx="50" cy="20" r="40" iid="0" class="item-normal"></circle> </g></svg></div>
    <app-upload-picture></app-upload-picture> <input type="file" accept="image/*" style="display: none;"></v-container>
`;
